jobs/e2e-test-cycloud.groovy

```diff
@@ -0,0 +1,230 @@
+   """IGNORE: This is a dummy parameter that is only here to avoid breaking the
+   communication with buildmaster""",
+   ""
+
+).apply();
+
+
+// We set these to real values first thing below; but we do it within
+// the notify() so if there's an error setting them we notify on slack.
+NUM_WORKER_MACHINES = null;
+// Used to tell whether all the test-workers raised an exception.
+WORKERS_RAISING_EXCEPTIONS = 0;
+public class TestFailed extends Exception {}  // for use with the above
+// Used to make sure we finish our job as soon as all tests are run.
+public class TestsAreDone extends Exception {}
+
+
+// Override the build name by the info that is passed in (from buildmaster).
+REVISION_DESCRIPTION = params.REVISION_DESCRIPTION ?: params.CYPRESS_GIT_REVISION;
+E2E_URL = params.URL[-1] == '/' ? params.URL.substring(0, params.URL.length() - 1): params.URL;
+
+currentBuild.displayName = ("${currentBuild.displayName} " +
+                            "(${REVISION_DESCRIPTION})");
+
+// We use the build name as a unique identifier for user notifications.
+BUILD_NAME = "build e2e-cypress-test #${env.BUILD_NUMBER} (${E2E_URL}: ${params.REVISION_DESCRIPTION})"
+
+// GIT_SHA1 is the sha1 for CYPRESS_GIT_REVISION.
+GIT_SHA1 = null;
+
+// We have a dedicated set of workers for the second smoke test.
+WORKER_TYPE = (params.USE_FIRSTINQUEUE_WORKERS
+               ? 'ka-firstinqueue-ec2' : 'ka-test-ec2');
+
+def initializeGlobals() {
+   NUM_WORKER_MACHINES = params.NUM_WORKER_MACHINES.toInteger();
+}
+
+
+def _setupWebapp() {
+   GIT_SHA1 = kaGit.resolveCommitish("git@github.com:Khan/webapp",
+                                        params.CYPRESS_GIT_REVISION);
+
+   kaGit.safeSyncToOrigin("git@github.com:Khan/webapp", GIT_SHA1);
+
+   dir("webapp/services/static") {
+      sh("yarn install --frozen-lockfile");
+   }
+}
+
+// Run all the test-clients on all the worker machine, in parallel.
+def runAllTestClients() {
+   // We want to swallow any framework exceptions unless *all* the
+   // clients have raised a framework exception.  Our theory is that
+   // if one client dies unexpectedly the others can compensate, but
+   // if they all do, then there's nothing more we can do.
+   def onException = {
+      echo("Worker raised an exception");
+      WORKERS_RAISING_EXCEPTIONS++;
+      if (WORKERS_RAISING_EXCEPTIONS == NUM_WORKER_MACHINES) {
+         echo("All worker machines failed!");
+         throw new TestFailed("All worker machines failed!");
+      }
+   }
+
+   def jobs = [:];
+   for (i = 0; i < NUM_WORKER_MACHINES; i++) {
+      def workerId = i;  // avoid scoping problems
+      jobs["e2e-test-${workerId}"] = {
+         swallowExceptions({
+            onWorker(WORKER_TYPE, '2h') {
+                _setupWebapp();
+                runE2ETests(workerId);
+            }
+         }, onException);
+      };
+   }
+   parallel(jobs);
+}
+
+
+def runTests() {
+   try {
+     runAllTestClients();
+    } catch (TestsAreDone e) {
```

Comment:
You don't make use of this -- you never throw it -- and can just remove it.  Likewise TestFailed.

Reply:
Thanks! I've removed that logic (including the failures part).